server/src-test/Hasura/Server/MigrateSpec.hs

```diff
@@ -64,20 +67,37 @@ spec
 spec pgConnInfo = do
   let dropAndInit time = CacheRefT $ flip modifyMVar \_ ->
         dropCatalog *> (swap <$> migrateCatalog time)
-
+      
   describe "migrateCatalog" $ do
     it "initializes the catalog" $ singleTransaction do
       (dropAndInit =<< liftIO getCurrentTime) `shouldReturn` MRInitialized
 
     it "is idempotent" \(NT transact) -> do
-      let dumpSchema = transact $
-            execPGDump (PGDumpReqBody ["--schema-only"] (Just False)) pgConnInfo
+      let dumpSchema = execPGDump (PGDumpReqBody ["--schema-only"] (Just False)) pgConnInfo
       time <- getCurrentTime
       transact (dropAndInit time) `shouldReturn` MRInitialized
-      firstDump <- dumpSchema
+      firstDump <- transact dumpSchema
       transact (dropAndInit time) `shouldReturn` MRInitialized
-      secondDump <- dumpSchema
+      secondDump <- transact dumpSchema
       secondDump `shouldBe` firstDump
+      
+    it "supports upgrades after downgrade to version 12" \(NT transact) -> do
+      let downgradeTo v = downgradeCatalog DowngradeOptions{ dgoDryRun = False, dgoTargetVersion = v }
+          upgradeToLatest time = CacheRefT $ flip modifyMVar \_ ->
+            swap <$> migrateCatalog time
+      time <- getCurrentTime
+      transact (dropAndInit time) `shouldReturn` MRInitialized
+      downgradeResult <- (transact . lift) (downgradeTo "12" time)
+      downgradeResult `shouldSatisfy` \case
+        MRMigrated{} -> True
+        _ -> False
+      transact (upgradeToLatest time) `shouldReturn` MRMigrated "12"
+      
+    it "supports downgrades for every Git tag" $ singleTransaction do
```

Comment:
@0x777 The test is now implemented, but I'm concerned that this will cause test failures when we tag a new release. Perhaps we should omit the latest tag? We wouldn't ever need to downgrade to the latest tag anyway, except for in development.

Reply:
Oh right, nice catch. We can omit the latest tag, it makes sense, when we are releasing `v1.2.0`, we only need to ensure that downgrade option exists to versions < `v1.2.0`. Now say we are releasing `v1.3.0` and say we forgot to add the downgrade option to `v1.2.0`, this test will fail when we are actually making the release, so we'll have to push a new commit, delete the old tag, add it at the new commit - it is a little inconvenient but I guess it is better than not having it. So immediately after making a release, someone opens a PR with the new mapping? Is that the suggested workflow?

@lexi-lambda thoughts?

Reply:
I think perhaps the best option is to check if the current version is itself a tag. If so, check for all but the latest tag (all but the current release), and if not, check all tags. This way, the build will not fail for the tag itself, but for any commits after it. Ideally the first commit after the tag can add the downgrade option.